tensorboard/webapp/notification_center/_data_source/types.ts

```diff
@@ -0,0 +1,34 @@
+/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+import {Observable} from 'rxjs';
+import {Notification} from '../_redux/notification_center_types';
+
+export interface NotificationCenterSuccessfulResponse {
+  notifications: Notification[];
+  error?: undefined;
+}
+
+export interface NotificationCenterFailedResponse {
+  notifications?: Notification[];
+  error: string;
+}
+
+export type NotificationCenterResponse =
+  | NotificationCenterSuccessfulResponse
+  | NotificationCenterFailedResponse;
```

Comment:
these two types combined in L28-30 feels very weird; 

I think it makes more sense to have:
```ts
export interface NotificationCenterSuccessfulResponse {
  notifications: Notification[];
}

export interface NotificationCenterFailedResponse {
  error: string;
}

export type NotificationCenterResponse =
  | NotificationCenterSuccessfulResponse
  | NotificationCenterFailedResponse;
```

but at the same time, wouldn't failures actually throw an error instead? (i.e., status code is not 200s-300s?). In that case, rxjs/http_client actually throws and there is a rxjs way of catching that (https://www.learnrxjs.io/learn-rxjs/operators/error_handling/catch; you've already done that in effects). If my intuition is correct, you will not need the error type.

Reply:
My initial thought was as your codeblock above and some type mismatching in reducer while accessing ".notifications" 
```
Property 'notifications' does not exist on type 'NotificationCenterResponse'.
Property 'notifications' does not exist on type 'NotificationCenterFailedResponse'.
 ```

For now I've removed the success/failure and just `NotificationCenterResponse`